File: csharp/behavioral/mediator.cs
namespace Gof.ModernCSharp.Behavioral;

// Mediator centralizes object communication so participants do not call each other directly.
// Modern C# note: often appears as a message bus, controller, queue, or application
// service. Avoid letting it become a dumping ground.
public static class Mediator
{
    sealed class ChatRoom
    {
        readonly List<User> _users = [];
        public void Join(User user) { _users.Add(user); user.Room = this; }
        public IEnumerable<string> Send(User sender, string message) =>
            _users.Where(user => user != sender).Select(user => user.Receive(sender.Name, message));
    }

    sealed class User(string name)
    {
        public string Name => name;
        public ChatRoom? Room { get; set; }
        public IEnumerable<string> Send(string message) => Room!.Send(this, message);
        public string Receive(string sender, string message) => $"{Name} got '{message}' from {sender}";
    }

    public static string Run()
    {
        var room = new ChatRoom();
        var ada = new User("Ada");
        room.Join(ada);
        room.Join(new User("Grace"));
        return string.Join(", ", ada.Send("hello"));
    }
}

File: py/behavioral/mediator.py
# Mediator centralizes object communication so participants do not call each other directly.
# Modern Python note: event buses, controllers, queues, and service layers often play
# this role. Avoid a mediator that becomes a dumping ground for unrelated behavior.
class ChatRoom:
    def __init__(self):
        self.users = []

    def join(self, user):
        self.users.append(user)
        user.room = self

    def send(self, sender, message):
        return [user.receive(sender.name, message) for user in self.users if user != sender]


class User:
    def __init__(self, name):
        self.name = name
        self.room = None

    def send(self, message):
        return self.room.send(self, message)

    def receive(self, sender, message):
        return f"{self.name} got '{message}' from {sender}"


if __name__ == "__main__":
    room = ChatRoom()
    ada = User("Ada")
    grace = User("Grace")
    room.join(ada)
    room.join(grace)
    print(ada.send("hello"))
